src/compute/src/render/mod.rs

```diff
@@ -269,31 +270,43 @@ pub fn build_compute_dataflow<A: Allocate>(
-                        Id::Global(object.id),
-                        CollectionBundle::from_collections(oks_v.clone(), err_v.clone()),
-                    );
-                    variables.insert(object.id, (oks_v, err_v));
-                }
+                let mut any_letrec = false;
                 for object in dataflow.objects_to_build {
-                    let id = object.id;
-                    let bundle = context.render_plan(object.plan, region, region.index());
-                    // We need to ensure that the raw collection exists, but do not have enough information
-                    // here to cause that to happen.
-                    let (oks, err) = bundle.collection.clone().unwrap();
-                    context.insert_id(Id::Global(object.id), bundle);
-                    let (oks_v, err_v) = variables.remove(&id).unwrap();
-                    oks_v.set(&oks);
-                    err_v.set(&err);
+                    if let Plan::LetRec { ids, values, body } = object.plan {
+                        assert!(!any_letrec, "Cannot render multiple instances of LetRec");
+                        any_letrec = true;
+                        // Build declared objects.
+                        // It is important that we only use the `Variable` until the object is bound.
+                        // At that point, all subsequent uses should have access to the object itself.
+                        let mut variables = BTreeMap::new();
+                        for id in ids.iter() {
+                            use differential_dataflow::operators::iterate::Variable;
+
+                            let oks_v = Variable::new(region, Product::new(Default::default(), 1));
+                            let err_v = Variable::new(region, Product::new(Default::default(), 1));
+
+                            context.insert_id(
+                                Id::Local(*id),
+                                CollectionBundle::from_collections(oks_v.clone(), err_v.clone()),
+                            );
+                            variables.insert(Id::Local(*id), (oks_v, err_v));
+                        }
+                        for (id, value) in ids.into_iter().zip(values.into_iter()) {
+                            let bundle = context.render_plan(value, region, region.index());
+                            // We need to ensure that the raw collection exists, but do not have enough information
+                            // here to cause that to happen.
+                            let (oks, err) = bundle.collection.clone().unwrap();
+                            context.insert_id(Id::Local(id), bundle);
+                            let (oks_v, err_v) = variables.remove(&Id::Local(id)).unwrap();
+                            oks_v.set(&oks);
+                            err_v.set(&err);
+                        }
+
+                        let bundle = context.render_plan(*body, region, region.index());
```

Comment:
This fragment seems to be a candidate to be moved into `Context::render_plan` under the `LetRec` rendering arm? Then, the code here would simplify to: (a) applying predicates over the objects to ensure constraints on the current rendering of recursive LIR expressions are satisfied (e.g., as you did inline with `any_letrec`); (b) calling `context.build_object` in each object (recursive or not).

Reply:
As far as I can tell, this cannot happen without either type shenanigans, or passing along a "should explode" bit and considering rendering stages that are nowhere near correct (e.g. ones that have the wrong timestamp types), or maybe both. I'm not sure which approach is better. For example, it would be hard to move these lines (and things related to their types) into `build_object`:
```rust
let oks_v = Variable::new(region, Product::new(Default::default(), 1));
let err_v = Variable::new(region, Product::new(Default::default(), 1));
```
because the `Product` timestamp type doesn't exist in that method.

Reply:
Got it, that makes sense, thanks!

Reply:
I think we should find a way to make this code better abstracted in the future, but at the moment I'm ok with it being a bit of a mess as we exercise it. If needed, we can function it up and figure out what the complicated signature should be.

Reply:
Sounds good, let's keep experimenting and moving forward for now. We may wish in the future to revisit, perhaps abstract a separate `IterativeContext`... anyway, let's see how the code shapes up as more progress is made!

Reply:
Yup, this might make sense. The types do tell us which of the two they are (if the `T` argument is `mz_repr::Timestamp` or something else), and providing methods only on the struct for which the timestamp is exactly that makes sense. I just tried writing some implementations on the iterative context (where the scope and timestamp are specific things) and .. Rust struggled to reach the same conclusions I reached about what was obviously true about some of the types, and we agreed to disagree (I gave up).